Add an enhancement proposal for initial ScyllaDBCluster integration with ScyllaDB Manager (#2558)
scylladb/scylla-operator: enhancements/proposals/2549-scylladbcluster-backup-repair/README.md

# ScyllaDBCluster integration with ScyllaDB Manager: cluster registration, backup and repair scheduling

## Summary

This proposal aims at introducing initial support for ScyllaDBCluster integration with ScyllaDB Manager.
The initial integration covers support for registration of managed, multi-datacenter clusters, and scheduling backup and repair tasks.

## Motivation

We want to introduce ScyllaDB Manager integration with ScyllaDBCluster to extend our existing integration to automated multi-datacenter clusters.
We want to support all the operations we currently support with single-datacenter clusters to at least the same extent.
This also allows us to deal with some of the issues and inconsistencies existing in ScyllaCluster's API related to ScyllaDB Manager integration.

### Goals

- Provide a suggested deployment mechanism for unmanaged ScyllaDB Manager server deployment in multi-datacenter Kubernetes environments.
- Introduce support for registration of ScyllaDBClusters with ScyllaDB Manager.

> - Introduce support for scheduling of backup tasks for ScyllaDBClusters with ScyllaDB Manager.
> - Introduce support for scheduling of repair tasks for ScyllaDBClusters with ScyllaDB Manager.
> - Maintain support for registering ScyllaClusters and scheduling backup and repair tasks for ScyllaClusters with ScyllaDB Manager.

Reviewer: Product-wise we're not "introducing support" but extending support to multi-DC clusters.
While this sentence is technically correct (because _technically_) we're introducing support to Scylla**DB**Clusters that was present in ScyllaClusters, this phrasing misses the practical sense, especially for outside readers.

Suggested (hopefully more accurate/useful) phrasing could be:

```suggestion
- Rework the CRD API for backup & repair tasks to decouple from the legacy `ScyllaCluster` CRD.
    - Includes a compatibility solution for legacy `ScyllaCluster` setups using the in-CRD backup/repair stanzas.
- Introduce a way to federate backup&repair tasks across Kubernetes clusters.
```

If you think it doesn't capture the meaning, feel free to correct me!

#### Definition of `ScyllaDBManagerTask.scylla.scylladb.com/v1alpha1`

```go
type ScyllaDBManagerTaskType string

const (
	ScyllaDBManagerTaskTypeBackup ScyllaDBManagerTaskType = "Backup"
	ScyllaDBManagerTaskTypeRepair ScyllaDBManagerTaskType = "Repair"
)

type ScyllaDBManagerTaskSchedule struct {
	// cron specifies the task schedule as a cron expression.
	// It supports the "standard" cron syntax `MIN HOUR DOM MON DOW`, as used by the Linux utility, as well as a set of non-standard macros: "@yearly", "@annually", "@monthly", "@weekly", "@daily", "@midnight", "@hourly", "@every [+-]?<duration>".
	// +optional
	Cron *string `json:"cron,omitempty"`

	// numRetries specifies how many times a scheduled task should be retried before failing.
	// +optional
	NumRetries *int64 `json:"numRetries,omitempty"`

	// startDate specifies the start date of the task.
	// If not set, the task is started immediately.
	// +optional
	StartDate *time.Time `json:"startDate,omitempty"`
}

type ScyllaDBManagerBackupTaskOptions struct {
	// schedule specifies the schedule on which the backup task is run.

> 	ScyllaDBManagerTaskSchedule `json:",inline"`

Reviewer 2: nit: maybe it wold make more sense to inline schedule into the `ScyllaDBManagerTaskSpec` instead of all tasks options separately?
It would also correspond closer to SM API, where the task is defined as:
```json
    "Task": {
      "type": "object",
      "properties": {
        ...
        "schedule": {
          "$ref": "#/definitions/Schedule"
        },
        "properties": {
          "type": "object",
          "additionalProperties": true
        }
      }
    },
```

Author: It would in terms of aligning with manager API, but it would also force us to handle it conditionally for a task type. It doesn't seem to make much sense with a restore, which is a one-time action, and that we plan to add in the future. Or are there any scenarios where restore on a schedule would make sense?

Reviewer 2: The only schedule option which makes any sense with restore is `--start-date`, but it's still not that useful, so it makes sense to have schedule outside of task spec.

A new set of controllers will be responsible for registering ScyllaDBClusters and scheduling tasks with ScyllaDB Manager.

The new API will allow for referencing both ScyllaDBClusters (scylla.scylladb.com/v1alpha1) and ScyllaClusters (scylla.scylladb.com/v1).
To support the legacy API without duplicating the reconciliation logic, the existing controller responsible for ScyllaCluster integration with ScyllaDB Manager will translate tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` into the new API objects.
Simultaneously, `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` will be deprecated and users advised to use the new API.

### User Stories

#### Unmanaged deployment of ScyllaDB Manager server and manager-controller in multi-datacenter environments

As a user, I want to deploy ScyllaDB Manager server deployment, together with the manager-controller deployment, in a multi-datacenter environment, to work with automated, multi-datacenter ScyllaDB clusters.

> To do that, I deploy ScyllaDB Manager server and manager-controller in a single datacenter, alongside the ScyllaDBDatacenter object.

Author: done

As a user, I want to register multi-datacenter ScyllaDB Clusters, reflected in Kubernetes by ScyllaDBClusters (scylla.scylladb.com/v1alpha1), with ScyllaDB Manager.
To do that, I mark the ScyllaDBCluster object with an appropriate label. 

#### ScyllaDB Manager backup and repair task scheduling for multi-datacenter ScyllaDB clusters

As a user, I want to schedule backup and repair tasks for multi-datacenter ScyllaDB clusters with ScyllaDB Manager.
To do that, I create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a ScyllaDBCluster.

#### ScyllaDB Manager backup and repair task scheduling for single-datacenter ScyllaDB clusters

As a user, I want to maintain the possibility of scheduling backup and repair tasks for single-datacenter ScyllaDB clusters with ScyllaDB Manager.
To do that, I either:
- Create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a single-datacenter ScyllaDBCluster.
- Configure the legacy ScyllaCluster (scylla.scylladb.com/v1) object with deprecated `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs`/

> - Create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a ScyllaCluster.

Author: > Shouldn't we support only ScyllaDBCluster and ScyllaDBDatacenter reference within ScyllaDBManagerTask?

Wouldn't this only make it more convoluted? The users won't be able to use ScyllaDBManagerTasks with ScyllaDBDatacenter directly by design - otherwise we could have conflicts with manager-controller registering the same cluster by reconciling both a ScyllaDBCluster and ScyllaDBDatacenter. Supporting ScyllaDBManagerTask with ScyllaClusters instead also allows us to deprecate it together with the ScyllaClusters.
Would a comment saying this is for backwards compatibility only and will be deprecated help here?

> - Create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a ScyllaCluster.

Author: Ok, it makes sense. I'll amend the document.

> - Create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a ScyllaCluster.

Author: changed, ptal

The existing manager-controller application will be extended with a dedicated controller reconciling ScyllaDBClusters.
The new controller will be responsible for registering ScyllaDBClusters with ScyllaDB Manager server.
To mark ScyllaDBClusters for registration with ScyllaDB Manager, users will label ScyllaDBClusters with a dedicated label: `scylla-operator.scylladb.com/register-with-scylladb-manager`.
The dedicated label serves as an intermediate step before we introduce managed ScyllaDB Manager server deployments, when the dedicated label will no longer be required as the dedicated API object will be able to select ScyllaDBClusters using the standard labels.
We do not modify ScyllaDBCluster API for this purpose. To save the identification number internal to ScyllaDB Manager state in Kubernetes state, a new, internal annotation is used: `internal.scylla-operator.scylladb.com/scylladb-manager-clusterid`.

For backwards compatibility, the existing controller, reconciling ScyllaClusters, keeps reconciling all ScyllaClusters in the Kubernetes cluster in which the application and ScyllaDB Manager server deployment are deployed.

#### Auth tokens

> ScyllaDBCluster controller, running as part of the operator, will generate the auth token secret and configure the created ScyllaDBDatacenters with an internal annotation `internal.scylla-operator.scylladb.com/sylladb-manager-agent-auth-token-override-secret-ref`.

Reviewer 3: ```suggestion
ScyllaDBCluster controller, running as part of the operator, will generate the auth token secret and configure the created ScyllaDBDatacenters with an internal annotation `internal.scylla-operator.scylladb.com/scylladb-manager-agent-auth-token-override-secret-ref`.
```

Author: done

    
    // type specifies the type of the task.
    Type ScyllaDBManagerTaskType `json:"type"`
    
    // backup specifies the options for a backup task.
    // +optional
    Backup *ScyllaDBManagerBackupTaskOptions `json:"backup,omitempty"`
    
    // repair specifies the options for a repair task.
    // +optional
    Repair *ScyllaDBManagerRepairTaskOptions `json:"repair,omitempty"`
}

type ScyllaDBManagerTaskStatus struct {
	// observedGeneration is the most recent generation observed for this ScyllaDBManagerTask. It corresponds to the
	// ScyllaDBManagerTask's generation, which is updated on mutation by the API Server.
	// +optional
	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`

	// conditions hold conditions describing ScyllaDBManagerTask state.
	// +optional
	Conditions []metav1.Condition `json:"conditions,omitempty"`

	// taskID reflects the internal identification number of the task in ScyllaDB Manager state.
	// +optional
	TaskID *string `json:"taskID,omitempty"`
}

// +kubebuilder:object:root=true
// +kubebuilder:subresource:status
// +kubebuilder:storageversion
// +genclient
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

> // +kubebuilder:printcolumn:name="AVAILABLE",type=string,JSONPath=".status.conditions[?(@.type=='Available')].status"

Reviewer 3: Available doesn't make sense in terms of ScyllaDBManagerTask, I would recommend to not print nor set it anywhere.

Author: done

> // +kubebuilder:printcolumn:name="PROGRESSING",type=string,JSONPath=".status.conditions[?(@.type=='Progressing')].status"

Reviewer 3: Same here, once generation == observedGeneration task can either be scheduled - and have non-nil TaskID - or be degraded.

Author: done

	// status reflects the observed state of ScyllaDBManagerTask.
	Status ScyllaDBManagerTaskStatus `json:"status,omitempty"`
}
```

#### Notable changes between `ScyllaDBManagerTask.scylla.scylladb.com/v1alpha1` and legacy `ScyllaClusters.scylla.scylladb.com/v1` and the conversion mechanisms

##### Status

The new API drops all status propagation but the internal identification number of the given task in ScyllaDB Manager state, also referred to as TaskID.
This is due to the fact that the new ScyllaDBCluster and ScyllaDBManagerTask controllers will no longer use the objects' statuses for performing updates.

The dedicated controller will propagate any errors returned by the ScyllaDB Manager client to the object's status conditions.

To maintain backwards compatibility, we keep reflecting the tasks' state in ScyllaDB Manager in `scyllaclusters.scylla.scylladb.com/v1.status.backups` and `scyllaclusters.scylla.scylladb.com/v1.status.repairs`.
The new ScyllaDBManagerTask controller will update ScyllaDBManagerTask objects with an internal annotation `internal.scylla-operator.scylladb.com/sylladb-manager-task-status` in respective JSON encoded structs.

> The translating controller will use the annotation values to update the ScyllaCluster

Reviewer 3: ```suggestion
The translating controller will use the annotation values to update the ScyllaCluster status.
```

Author: done

When referencing a ScyllaDBCluster, ScyllaDBManagerTask objects created by the user will not be garbage collected on ScyllaDBCluster deletion. Users will be responsible for deleting the objects.

#### Providing credentials for object storage access by ScyllaDB Manager agent sidecar

ScyllaDB Manager agent containers require credentials to the corresponding regional object storage.
Users will be responsible for configuring the credentials for each datacenter by defining `scylladbclusters.scylla.scylladb.com/v1alpha1.spec.datacenters[].scyllaDBManagerAgent.volumes` and `scylladbclusters.scylla.scylladb.com/v1alpha1.spec.datacenters[].scyllaDBManagerAgent.volumeMounts`.

### Test Plan

The existing set of E2E tests related to ScyllaCluster integration with ScyllaDB Manager will be maintained in an unchanged form. 
The tests will cover correctness of ScyllaCluster's task spec conversion to ScyllaDBManagerTasks and ScyllaDBManagerTasks' statuses to ScyllaCluster's task statuses.

> Conversion logic between ScyllaCluster's (scylla.scylladb.com/v1) task specification and ScyllaDBManagerTasks (scylla.scylladb.com/v1alpha1) will also be covered by unit tests.

Reviewer 3: ```suggestion
Conversion logic between ScyllaCluster's (scylla.scylladb.com/v1) task specification and ScyllaDBManagerTasks (scylla.scylladb.com/v1alpha1) will be covered by unit tests.
```

Backup and repair tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` will no longer be reconciled directly by manager-controller.
Instead, the existing ScyllaCluster translation controller, running as part of the operator, is extended to convert the backup and repair tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` into the ScyllaDBManagerTask objects and apply them.

ScyllaDBManagerTask objects, created by the translation controller, adhere to the following naming convention to avoid name collisions:
- Tasks created from `scyllaclusters.scylla.scylladb.com/v1.spec.backups`: SHA256(`scyllaclusters.scylla.scylladb.com/v1.name`, "backup", `scyllaclusters.scylla.scylladb.com/v1.spec.backups[].name`).
- Tasks created from `scyllaclusters.scylla.scylladb.com/v1.spec.repairs`: SHA256(`<scyllaclusters.scylla.scylladb.com/v1.name>`, "repair", `scyllaclusters.scylla.scylladb.com/v1.spec.repairs[].name`).

> SHA256 hash is to reduce the length of the produced name, as both ScyllaCluster's and corresponding task's name could have max allowed name length. Hexadecimal representation of SHA256 is 64 characters.

Reviewer: Is the only reason why we wouldn't just apply a `backup-` or `restore-` prefix to a preexisting name that we might be exceeding the max length?

I'd be inclined to keep the generated names readable. And if collisions because of truncation are a concern, we can solve that by adding an alphanum stable suffix similar t ohttps://github.com/scylladb/scylla-operator/pull/2505.

Author: > Is the only reason why we wouldn't just apply a backup- or restore- prefix to a preexisting name that we might be exceeding the max length?

Yes.

I suppose something like below is also an option, with the risk of exceeding max length, which is a bit silly of a reason to error on reconciliation imo. It's an under-the-hood object when created in ScyllaCluster translation, so is readability really a concern?

```
		nameSuffix, err := naming.GenerateNameHash(string(scyllav1alpha1.ScyllaDBManagerTaskTypeBackup), backupSpec.Name)
		if err != nil {
			return progressingConditions, fmt.Errorf("can't generate ScyllaDBManagerTask name suffix: %w", err)
		}
		name := fmt.Sprintf("%s-%s", sc.Name, nameSuffix)
```

Author: done

- Propagate manager-controller status conditions to ScyllaDBCluster or ScyllaCluster [#1921][3].

## Proposal

I propose to introduce support for ScyllaDBClusters' (scylla.scylladb.com/v1alpha1) and ScyllaDBDatacenters' (scylla.scylladb.com/v1alpha1) registration with ScyllaDBManager with no changes to ScyllaDBCluster nor ScyllaDBDatacenter API.
For this measure, I propose adding a new API object: `ScyllaDBManagerClusterRegistration.scylla.scylladb.com/v1alpha1`. Other than keeping the metadata involved in cluster registration with ScyllaDB Manager, the purpose of the intermediate resource is future-proofing the approach for introduction of managed ScyllaDB Manager instances, in which case the new API can also be extended to serve as a reverse reference to a ScyllaDB Manager instance.
It is an internal resource managed by Scylla Operator and users are not expected to perform any operations involving it.

To support task scheduling, I propose to add a new API object: `ScyllaDBManagerTask.scylla.scylladb.com/v1alpha1`.
The new API will allow for referencing both ScyllaDBClusters (scylla.scylladb.com/v1alpha1) and ScyllaDBDatacenters (scylla.scylladb.com/v1alpha1).

> Manager-controller, today deployed alongside ScyllaDB Manager server, will be removed in favour of a new set of controllers, extending Scylla Operator, responsible for registering ScyllaDB clusters and scheduling tasks with unmanaged ScyllaDB Manager server.

Author: done

> To support the legacy API without duplicating the reconciliation logic, the existing controller responsible for ScyllaCluster migration will translate tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` into the new API objects.

Author: We agreed on not deprecating it as there won't be a direct alternative to scheduling tasks for ScyllaCluster.v1, only for the underlying ScyllaDBDatacenter.v1alpha1. `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` should be deprecated together with the entire ScyllaCluster.v1 resource.

> To support the legacy API without duplicating the reconciliation logic, the existing controller responsible for ScyllaCluster migration will translate tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` into the new API objects.

Reviewer: discussed offline - result as @rzetelskik described above.

As a user, I want to register multi-datacenter ScyllaDB clusters, reflected in Kubernetes by ScyllaDBClusters (scylla.scylladb.com/v1alpha1), with ScyllaDB Manager.
To do that, I mark the ScyllaDBCluster object with a predefined label. 

#### ScyllaDB Manager backup and repair task scheduling for multi-datacenter ScyllaDB clusters registered with ScyllaDB Manager

As a user, I want to schedule backup and repair tasks for multi-datacenter ScyllaDB clusters registered with ScyllaDB Manager.
To do that, I create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a ScyllaDBCluster.

#### ScyllaDB Manager backup and repair task scheduling for single-datacenter ScyllaDB clusters registered with ScyllaDB Manager

As a user, I want to maintain the possibility of scheduling backup and repair tasks for single-datacenter ScyllaDB clusters registered with ScyllaDB Manager.
To do that, I either:
- Create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a single-datacenter ScyllaDBCluster.
- Create ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) objects referencing a ScyllaDBDatacenter.

> - Configure the legacy ScyllaCluster (scylla.scylladb.com/v1) object with `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs`.

Author: ok, but changed deprecated to legacy

Clusters or tasks already existing in ScyllaDB Manager state should not be abandoned nor registered/scheduled multiple times.

To mitigate this, we override the naming convention for clusters and tasks defined with legacy ScyllaCluster (scylla.scylladb.com/v1) API.

#### Multiple ScyllaDB Manager server deployments reconciling the same cluster with a manual multi-datacenter procedure

Users following the manual procedure for deploying multi-datacenter ScyllaDB clusters in interconnected Kubernetes clusters (["Deploy a multi-datacenter ScyllaDB cluster in multiple interconnected Kubernetes clusters"][5]) may end up with multiple ScyllaDB Manager server instances registering and working with the cluster if they deploy ScyllaDB Manager in multiple Kubernetes clusters [#2119][6].
Users will be expected to deploy ScyllaDB Manager in only one datacenter and/or switch to the automated multi-datacenter ScyllaDB clusters.

## Design Details

### Deployment mechanism for ScyllaDB Manager

#### Changes to the deployment mechanism in single-datacenter environments

The existing deployment mechanism in single-datacenter environments is modified.

> The standalone manager-controller must no longer be deployed. Its functionality is replaced by a new set of controllers extending Scylla Operator.

Author: The upgrade procedure is described in a separate section, why duplicate it here?

To work with ScyllaDBClusters in multi-datacenter environments, ScyllaDB Manager server must be deployed in the "meta" cluster, in which the ScyllaDBCluster, that the ScyllaDB Manager server should integrate with, is deployed.
ScyllaDB Manager server must be deployed in `scylla-manager` namespace.
The standalone manager-controller must not be deployed.

### Registration of ScyllaDB clusters with ScyllaDB Manager

To support ScyllaDB cluster registration of both ScyllaDBClusters (scylla.scylladb.com/v1alpha1) and ScyllaDBDatacenters (scylla.scylladb.com/v1alpha1), we introduce a new API: `ScyllaDBManagerClusterRegistration.scylla.scylladb.com/v1alpha1`.
We extend Scylla Operator with a new controller: GlobalScyllaDBManager controller. The controller does not correspond to any existing Kubernetes resources but the "virtual" ScyllaDB Manager instance.
GlobalScyllaDBManager controller will create a ScyllaDBManagerClusterRegistration in the reserved `scylla-manager` namespace for all ScyllaDBCluster and ScyllaDBDatacenter objects with `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"` label across all namespaces.

> No other clusters in ScyllaDB Manager's state will be subject to reconciliation.

Author: I think it changes the meaning. What I wanted to describe here is that the controller won't reconcile any clusters in Manager's state that are not reflected in K8s API. Other clusters can be registered with ScyllaDB Manager, they just won't be reconciled by us.

Scylla Operator will be extended with a controller reconciling ScyllaDBManagerClusterRegistration, responsible for registering ScyllaDBDatacenter and ScyllaDBCluster objects with a ScyllaDB Manager instance.
For the time being, only ScyllaDBManagerClusterRegistration objects with the internal annotation `"internal.scylla-operator.scylladb.com/global-scylladb-manager": "true"` will be reconciled.

To mark ScyllaDBClusters or ScyllaDBDatacenters for registration with the global ScyllaDB Manager server, users will label the objects with a dedicated label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`.
To avoid ScyllaDB manager registering the same cluster through both ScyllaDBCluster and the underlying ScyllaDBDatacenters, the ScyllaDBCluster controller will not propagate the label to the created ScyllaDBDatacenters.

#### Naming convention in ScyllaDB Manager state

ScyllaDBManagerClusterRegistration controller will set the name of the clusters registered in the global ScyllaDB Manager's state to `scylladbmanagerclusterregistration.scylla.scylladb.com/v1alpha1.metadata.name`.

> To ensure continuity with the existing state of global ScyllaDB Manager instance, we will enable overriding the name with an internal annotation `"internal.scylla-operator.scylladb.com/scylladb-manager-cluster-registration-name-override"`.

Author: `internal.scylla-operator.scylladb.com/scylladb-manager-cluster-registration-name-override` overrides a name under which `ScyllaDBManagerClusterRegistration` controller creates a cluster with global ScyllaDB Manager.

`internal.scylla-operator.scylladb.com/scylladbcluster-ref-namespace-override` overrides the implied local namespace in the reference to ScyllaDBCluster/ScyllaDBDatacenter, used by the controllers to `GET` the objects.

Imo the function is dissimilar and so should the annotation/label names in that case.

Would this make more sense?
`internal.scylla-operator.scylladb.com/scylladbmanagerclusterregistration-name-override`

`internal.scylla-operator.scylladb.com/scylladbmanagerclusterregistration-scylladbclusterref-namespace-override`

	// clusterID reflects the internal identification number of the cluster in ScyllaDB Manager state.
	// +optional
	ClusterID *string `json:"clusterID,omitempty"`
}

// +kubebuilder:object:root=true
// +kubebuilder:subresource:status
// +kubebuilder:storageversion
// +genclient
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
// +kubebuilder:printcolumn:name="DEGRADED",type=string,JSONPath=".status.conditions[?(@.type=='Degraded')].status"
// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"

type ScyllaDBManagerClusterRegistration struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec defines the desired state of ScyllaDBManagerClusterRegistration.
	Spec ScyllaDBManagerClusterRegistrationSpec `json:"spec,omitempty"`

	// status reflects the observed state of ScyllaDBManagerClusterRegistration.
	Status ScyllaDBManagerClusterRegistrationStatus `json:"status,omitempty"`
}
```

#### `ScyllaCluster.scylla.scylladb.com/v1` conversion mechanism

The existing ScyllaCluster (scylla.scylladb.com/v1) controller, running as part of the standalone manager-controller, will be removed.

> To ensure backwards compatibility and keep reconciling all ScyllaClusters in the Kubernetes cluster with the global ScyllaDB Manager instance, the migration controller by default labels all ScyllaDBDatacenters created as a result of migration from ScyllaClusters with the label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`. 

Author: > what makes annotations/labels internal vs not internal?

internal - meant to be set by operator (implied higher degree of flexibility for us imo)
non-internal - meant to be set by user

> we can distinguish between "public" and "do not use" in documentation, and the "internal"ness may not survive the test of time.

We don't really document the annotations/labels and we can always make an internal one non-internal?

> To ensure backwards compatibility and keep reconciling all ScyllaClusters in the Kubernetes cluster with the global ScyllaDB Manager instance, the migration controller by default labels all ScyllaDBDatacenters created as a result of migration from ScyllaClusters with the label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`. 

Reviewer: I'm coming from an assumption that the intention (to set by operator or by the user) can evolve over time (especially as e.g. bugs get reported and we need workarounds by setting them manually) - and then the `internal` factor will inevitably lose its meaning.

Therefore I see "internal"/non-"internal" as a not very useful category and suggest dropping the distinction on the name level.

This is a weakly held opinion - feel free to decide on your own :+1:

> To ensure backwards compatibility and keep reconciling all ScyllaClusters in the Kubernetes cluster with the global ScyllaDB Manager instance, the migration controller by default labels all ScyllaDBDatacenters created as a result of migration from ScyllaClusters with the label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`. 

Author: @zimnx do you want to chime in?

> To ensure backwards compatibility and keep reconciling all ScyllaClusters in the Kubernetes cluster with the global ScyllaDB Manager instance, the migration controller by default labels all ScyllaDBDatacenters created as a result of migration from ScyllaClusters with the label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`. 

Reviewer: Right -- difference in support commitment between internal and non-internal sounds convincing to me. :+1:

    ScyllaDBManagerTask controller will use it to override the value set in the object's specification.
  - The default value for `FailFast` is removed from the new API. The ScyllaDB Manager server assumes a default value if the parameter is no set explicitly.
  - The default value for `SmallTableThreshold` is removed from the new API. The ScyllaDB Manager server assumes a default value if the parameter is no set explicitly.
  - The default value for `Retention` is removed from the new API. The ScyllaDB Manager server assumes a default value if the parameter is no set explicitly.
  - The default value for `Parallel` is removed from the new API. The ScyllaDB Manager server assumes a default value if the parameter is no set explicitly.
  - ScyllaDBManagerTask controller will no longer override `NumRetries` when `FailFast` is set to true. If required, such behaviour should be enforced by ScyllaDB Manager server.
    The translating controller will maintain this behaviour when applying ScyllaDBManagerTasks.

The translating controller will keep respecting the default values defined in ScyllaCluster API and propagate them to applied ScyllaDBManagerTasks.

#### Finalizers

> The new ScyllaDBManagerTask controller will set a `scylla-operator.scylladb.com/scylladbmanagertask-protection` finalizer on ScyllaDBManagerTask objects so that the tasks are deleted from ScyllaDB Manager state on object deletion.

Reviewer: "protection" will mean nothing to the user. If deletion gets stuck on "protection", they'll be absolutely clueless.

I suggest using a name that will allow the user to infer the meaning. For example:

```suggestion
The new ScyllaDBManagerTask controller will set a `scylla-operator.scylladb.com/scylladbmanagertask-ensure-deregistration` finalizer on ScyllaDBManagerTask objects so that the tasks are deleted from ScyllaDB Manager state on object deletion.
```

Author: I was trying to stay consistent with existing RemoteKubernetesCluster and ScyllaDBCluster finalizers, but that's a fair point.
Would `scylla-operator.scylladb.com/scylladbmanagertask-deregistration` suffice though?

Reviewer: absolutely :+1: thanks!

Author: changed to `scylla-operator.scylladb.com/scylladbmanagertask-deletion`, as that's the operation performed with manager

#### Status conditions

The new ScyllaDBManagerTask controller will propagate ScyllaDB Manager client errors to `Degraded` status condition.

The translating controller will propagate the error to a corresponding task's status in ScyllaCluster object.

#### Owner references

The translating controller will set the owner references of the created ScyllaDBManagerTask objects to a corresponding ScyllaCluster. The created ScyllaDBManagerTask objects will be garbage collected on ScyllaCluster's deletion according to the deletion policy.

#### Providing credentials for object storage access by ScyllaDB Manager agent sidecar

ScyllaDB Manager agent containers require credentials to the corresponding regional object storage.

> Users will be responsible for configuring the credentials for each datacenter by defining `scylladbclusters.scylla.scylladb.com/v1alpha1.spec.datacenters[].scyllaDBManagerAgent.volumes` and `scylladbclusters.scylla.scylladb.com/v1alpha1.spec.datacenters[].scyllaDBManagerAgent.volumeMounts`.

Reviewer: (not actionable, just yelling into the void)

the fact that we're speaking in terms of container volumes makes this API a leaky abstraction :cry:. I understand that this is outside of this design's scope and we can't do anything about it in this design doc.

Author: > I understand that this is outside of this design's scope and we can't do anything about it in this design doc.

We could but only by heavily interfering with ScyllaDBCluster/ScyllaDBDatacenter API imo, while it was one of the assumptions that we don't touch it for manager-related operations.

Users will have to return to using task specifications in ScyllaCluster (scylla.scylladb.com/v1). The new CRs will not be migrated back. 
The new CR objects will be left behind and users might have to delete them manually from Kubernetes, as well as remove their corresponding objects from ScyllaDB Manager state.

### Version Skew Strategy

#### New Scylla Operator with current global ScyllaDB Manager deployment mechanism (manager-controller not removed)

The new set of controllers extending Scylla Operator will register each ScyllaCluster (scylla.scylladb.com/v1) and schedule its specified tasks with the global ScyllaDB Manager instance.
Simultaneously, the not removed manager-controller will keep reconciling ScyllaClusters as is, leading to conflicts.
Since we will retain the naming convention for ScyllaClusters registered with ScyllaDB Manager, as well as their scheduled tasks, the conflict should not lead to creating multiple cluster or task instances in ScyllaDB Manager state.
Minor changes in cluster or task specification, used for managed hash calculation in either controller, might cause the controllers to end up in an update hot loop.

> We can consider implementing a best-effort approach to preventing this scenario by refusing to start Scylla Operator if manager-controller deployment is discovered in `scylla-manager`.

Reviewer: I suggest elevating this to a requirement. Considering that:
- The hot loop is bad enough, and 
- the cost of the validation (including the need to remove it later)

the cost to benefit ratio appears strongly in favor of implementing the validation.

Reviewer 3: > preventing this scenario by refusing to start Scylla Operator if manager-controller deployment is discovered in `scylla-manager`.

This is quite harsh. Our main goal is to deploy ScyllaDB, often mission critical database. Misconfiguration of Scylla Manager shouldn't influence ability to reconcile clusters. 
We can stop particular controller (the one doing stuff around manager) in this case and make noise about it in logs, but not entire app.

Author: > We can stop particular controller (the one doing stuff around manager)

I think it would actually suffice if the migration controller:
a) didn't label the ScyllaDBDatacenter for registration with manager
b) didn't create ScyllaDBManagerTasks

That way all ScyllaClusters would still be managed by manager-controller and we'd avoid collisions, while the new features would still be available for ScyllaDBDatacenters and ScyllaDBClusters. 

Having said that, I think the logs will go unnoticed if the operator upgrade goes well. 

> This is quite harsh. Our main goal is to deploy ScyllaDB, often mission critical database. Misconfiguration of Scylla Manager shouldn't influence ability to reconcile clusters.

I think the assumption is that the operator outage after the upgrade won't really influence the running clusters. But I agree, this is quite a disruptive approach.

@mflendrich what do you think?

Reviewer: Discussed offline with @rzetelskik. I think that keeping manager-controller functional for ScyllaClusters while implementing the same responsibilities in operator (for SDBD/SDBC) is no good because it will be super confusing and (as @rzetelskik pointed out) will lead to some users relying on lingering manager-controller deployments when we're phasing it out.

My suggestion is that we stick with the design decision that running manager-controller is verboten, and we're only deciding if that should be a hard (=refuse to start up) or soft (=fall back to a "somewhat safe" state with warnings in logs) failure.

@zimnx two questions to see how we can reach some decision here.
1. In case of an upgrade in progress, do you agree (with my current view) that immediate feedback allows the SRE persona to react and correct immediately, and that leaving this only in logs is likely to cause them to miss the problem.
2. Do you share my view that this  "hard" validation is only an issue if the upgrade procedure is incorrectly executed, and will only cause failure to start up immediately after.

Based on the two assumptions above, I'm advocating that incorrect execution of the release procedure is a serious enough situation to get it corrected (and effectively disallow running in the undefined state).

Author: done

#### Current Scylla Operator with new global ScyllaDB Manager deployment mechanism (manager-controller removed)

No component will perform cluster registration or task scheduling with ScyllaDB Manager. ScyllaDB Manager state will not be lost.
The manager-controller will not reconcile the tasks defined in ScyllaCluster (scylla.scylladb.com/v1).

## Implementation History

- 2024-03-17: Initial enhancement proposal.
- 2024-04-02: Proposal extended with ScyllaDBManagerClusterRegistration (scylla.scylladb.com/v1alpha1) in an effort to future-proof the implementation for a move to managed ScyllaDB Manager deployments.

## Drawbacks

### No opt-out mechanism for ScyllaDB Manager integration

Introducing a controller of a "virtual" global ScyllaDB Manager instance comes with a drawback of no mechanism to opt-out of ScyllaDB Manager integration, up to this point achieved by not deploying the standalone manager-controller.

> Users can still opt out of using ScyllaDB Manager with Scylla Operator by not deploying the global ScyllaDB Manager server instance. We can consider checking for its presence in GlobalScyllaDBManager controller to avoid congesting the logs with misleading errors.

Reviewer: I suggest elevating this (a presence check) to a requirement, to keep the logic correct.

Author: I think making this a requirement, as well as the issues with ownership/finalisation call for deciding on what we consider a reflection of the global ScyllaDB manager in k8s state, since we don't have a single umbrella resource - is it a deployment, the backing ScyllaCluster, the service or the entire namespace? 

Also, especially important in terms of ScyllaDBManagerClusterRegistration finalisation - do we consider a deletion of the selected ScyllaDB Manager resource (deployment, SC, service, the entire namespace?) enough to consider the cluster as already deleted and make cluster deregistration a no-op in that case?

Reviewer: I think that we need to draw the line between "managed" and "unmanaged" manager somewhere. I feel that handling the events of "manager instance disappeared" and "manager instance appeared" comes very close to "managed manager".

And what we really need here is knowing whether the user intended to run Operator _with_ or_without_ the global manager. Autodetection based on the presence of a namespace sounds like a reasonable approach to me (the alternative I'd consider is a "hard" boolean flag like `--with-global-manager`). But let's not split hairs here and just go with autodetection based on the namespace's presence, IMO.

As for finalization

> do we consider a deletion of the selected ScyllaDB Manager resource (deployment, SC, service, the entire namespace?) enough to consider the cluster as already deleted and make cluster deregistration a no-op in that case?

I would rephrase it differently "deregistration is considered completed if the instance of manager we're trying to deregister in no longer exists".

Scylla Operator will be extended with a controller reconciling ScyllaDBManagerClusterRegistration, responsible for registering ScyllaDBDatacenter and ScyllaDBCluster objects with a ScyllaDB Manager instance.
For the time being, only ScyllaDBManagerClusterRegistration objects with the internal annotation `"internal.scylla-operator.scylladb.com/global-scylladb-manager": "true"` will be reconciled.

To mark ScyllaDBClusters or ScyllaDBDatacenters for registration with the global ScyllaDB Manager server, users will label the objects with a dedicated label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`.
To avoid ScyllaDB manager registering the same cluster through both ScyllaDBCluster and the underlying ScyllaDBDatacenters, the ScyllaDBCluster controller will not propagate the label to the created ScyllaDBDatacenters.

#### Naming convention in ScyllaDB Manager state

ScyllaDBManagerClusterRegistration controller will set the name of the clusters registered in the global ScyllaDB Manager's state to `scylladbmanagerclusterregistration.scylla.scylladb.com/v1alpha1.metadata.name`.

> To ensure continuity with the existing state of global ScyllaDB Manager instance, we will enable overriding the name with an internal annotation `"internal.scylla-operator.scylladb.com/name-override"`.

Author: good point, but for consistency I'd change it to `"internal.scylla-operator.scylladb.com/scylladb-manager-cluster-name-override"`, wdyt?

> To ensure continuity with the existing state of global ScyllaDB Manager instance, we will enable overriding the name with an internal annotation `"internal.scylla-operator.scylladb.com/name-override"`.

Reviewer 3: sounds good

> To ensure continuity with the existing state of global ScyllaDB Manager instance, we will enable overriding the name with an internal annotation `"internal.scylla-operator.scylladb.com/name-override"`.

Author: done

GlobalScyllaDBManager controller will propagate the annotation from ScyllaDBDatacenter and ScyllaDBCluster object to ScyllaDBManagerClusterRegistration objects.

#### Definition of `ScyllaDBManagerClusterRegistration.scylla.scylladb.com/v1alpha1`

```go
type LocalScyllaDBReference struct {
	// kind specifies the type of the resource.
	Kind string `json:"kind"`
	// name specifies the name of the resource.
	Name string `json:"name"`
}

type ScyllaDBManagerClusterRegistrationSpec struct {
	// scyllaDBClusterRef specifies the typed reference to the local ScyllaDB cluster.

> 	// Supported kinds are ScyllaDBCluster (scylla.scylladb.com/v1alpha1) and ScyllaDBDatacenter (scylla.scylladb.com/v1alpha1).

Author: done

	ScyllaDBClusterRef LocalScyllaDBReference `json:"scyllaDBClusterRef"`
}

type ScyllaDBManagerClusterRegistrationStatus struct {

> 	// clusterID reflects the internal identification number of the cluster in ScyllaDB Manager state.

Author: done

In order to set up E2E tests covering backup task scheduling for ScyllaDBClusters in multi-datacenter environments, the test-runner and related scripts will be extended to accept object storage bucket location and credentials for multiple datacenters.
The relevant CI job definitions will also be adjusted.
No changes are expected in the CI tooling.

### Upgrade / Downgrade Strategy

The new ScyllaDBManagerClusterRegistration (scylla.scylladb.com/v1alpha1) and ScyllaDBManagerTask (scylla.scylladb.com/v1alpha1) CRDs will have to be installed ahead of Scylla Operator upgrade. 
Our documentation and release notes cover this requirement for both GitOps and Helm installation paths on the regular basis.

As the existing manager-controller deployment is dropped with this proposal, users following the Helm installation path will have to upgrade their scylla-manager release ahead of Scylla Operator upgrade. 
Users following the GitOps installation path will have to manually delete the manager-controller deployment from `scylla-manager` namespace, and perform the regular update of ScyllaDB Manager manifests ahead of Scylla Operator upgrade.

> Release notes should explicitly mention and highlight the manual steps required on the GitOps path. 

Reviewer 3: Have you verified if Helm would remove the deployment upon upgrade? If it wouldn't we would have to mention it for Helm as well.

ScyllaDBManagerClusterRegistration (scylla.scylladb.com/v1alpha1) is introduced as a measure for Scylla Operator to maintain reverse references, allowing controllers to learn which ScyllaDB Manager instance a ScyllaDB cluster is registered with.
It is an internal resource managed by Scylla Operator and users are not expected to perform any operations involving it.

To support task scheduling, I propose to add a new API object: `ScyllaDBManagerTask.scylla.scylladb.com/v1alpha1`.

The new API will allow for referencing both ScyllaDBClusters (scylla.scylladb.com/v1alpha1) and ScyllaDBDatacenters (scylla.scylladb.com/v1alpha1).

Manager-controller, today deployed alongside ScyllaDB Manager server, will be removed in favour of a new set of controllers, extending Scylla Operator, responsible for registering ScyllaDBClusters and scheduling tasks with unmanaged ScyllaDB Manager server.

To support the legacy API without duplicating the reconciliation logic, the existing controller responsible for ScyllaCluster translation will translate tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` into the new API objects.

> Correspondingly, any 

Author: fixed

### User Stories

#### Unmanaged deployment of ScyllaDB Manager server in single-datacenter environments

> As a user, I want to deploy ScyllaDB Manager server deployment in a single-datacenter environment (single Kubernetes cluster), to work with automated, single-datacenter ScyllaDB clusters.

Author: I wrote this to highlight the changed deployment mechanism in single-datacenter environments. I'm not sure I understand your comment correctly - should this just not be mentioned as a user story if there already is a solution to this, even though the procedure changes with this proposal?

> As a user, I want to deploy ScyllaDB Manager server deployment in a single-datacenter environment (single Kubernetes cluster), to work with automated, single-datacenter ScyllaDB clusters.

Reviewer: I'm trying to say that what comes after _As a user, I want to_ does not express the desire of the user, and we're making a false statement that by changing the manager integration from one shape to another creates value for the user in the form of being able to deploy Manager.

Which is false for two reasons
- they already can do it, only using different means
- that does not provide value by itself.

> As a user, I want to deploy ScyllaDB Manager server deployment in a single-datacenter environment (single Kubernetes cluster), to work with automated, single-datacenter ScyllaDB clusters.

Author: fixed, hopefully

To do that, I deploy a predefined ScyllaDBManager object and ScyllaDB Manager server in `scylla-manager` namespace.

#### Unmanaged deployment of ScyllaDB Manager server in multi-datacenter environments

As a user, I want to deploy ScyllaDB Manager server deployment in a multi-datacenter environment, to work with automated, multi-datacenter ScyllaDB clusters.

> To do that, I deploy a predefined ScyllaDBManager object and ScyllaDB Manager server in `scylla-manager` namespace in a single Kubernetes cluster, in which the ScyllaDBCluster object is deployed.

Author: Again, I'm not sure what the ask here is. Do I understand correctly that you think the deployment user stories should be merged into the ones describing setting up tasks?

> To do that, I deploy a predefined ScyllaDBManager object and ScyllaDB Manager server in `scylla-manager` namespace in a single Kubernetes cluster, in which the ScyllaDBCluster object is deployed.

Reviewer: I'm trying to get us to use user stories as a tool to explore what value increment we're providing to the user (and where our blind spots are - like specific bits of documentation, for example). We shall take this offline.

We introduce a new API, `ScyllaDBManager.scylla.scylladb.com/v1alpha1`, which, at length, will serve as a declaration of intent of creating ScyllaDB Manager infrastructure and registering selected ScyllaDB cluster with ScyllaDB Manager server.
For the time being, however, the resource will only serve as a reflection of the unmanaged, global ScyllaDB Manager server deployed in `scylla-manager` namespace.

The managed ScyllaDB Manager infrastructure deployments will be a part of a separate effort, not described in this proposal.

#### Definition of `ScyllaDBManager.scylla.scylladb.com/v1alpha1`

```go
type LocalScyllaDBSelector struct {
	// kind specifies the type of the resource.
	Kind string `json:"kind"`

	// labelSelector specifies the label selector for resources of the specified kind.
	LabelSelector metav1.LabelSelector `json:"labelSelector"`
}

type ScyllaDBManagerSpec struct {
	// selectors specify which ScyllaDB clusters should be registered with ScyllaDBManager.
	// Supported kinds are ScyllaDBCluster (scylla.scylladb.com/v1alpha1) and ScyllaDBDatacenter (scylla.scylladb.com/v1alpha1).
	// A disjunction is used to combine all selectors.
	// +optional

> 	ScyllaDBClusterSelectors []LocalScyllaDBSelector `json:"scyllaDBClusterSelectors,omitempty"`

Author: > This leaves me with an impression of being overengineered: why have different selectors for different kinds? Why not a single selector that captures all the types supported?

That would lead to collisions between ScyllaDBClusters and underlying ScyllaDBDatacenters in the same namespace, as the labels are propagated to ScyllaDBDatacenter created by the ScyllaDBCluster controller. Although currently proposed API could also lead to it if users are not careful. What do you propose? Have a single selector for all supported types and programmatically exclude ScyllaDBDatacenters created by ScyllaDBCluster controller?

> 	ScyllaDBClusterSelectors []LocalScyllaDBSelector `json:"scyllaDBClusterSelectors,omitempty"`

Author: no longer applies (for now :wink:)

type ScyllaDBManager struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec defines the desired state of ScyllaDBManager.
	Spec ScyllaDBManagerSpec `json:"spec,omitempty"`

	// status reflects the observed state of ScyllaDBManager.
	Status ScyllaDBManagerStatus `json:"status,omitempty"`
}
```

#### Global instance of ScyllaDBManager (scylla.scylladb.com/v1alpha1)

Given we only aim to support ScyllaDBManager (scylla.scylladb.com/v1alpha1) as a reflection of a global, unmanaged instance of ScyllaDB Manager server, objects must have a predefined `scylla-operator.scylladb.com/global-scylladb-manager` annotation set to "true".
  
#### Deployment mechanism in single-datacenter environments

The existing deployment mechanism in single-datacenter environments is modified to include a ScyllaDBManager instance.
A single ScyllaDBManager instance must be deployed in `scylla-manager` namespace, with:
- A `scylla-operator.scylladb.com/global-scylladb-manager` annotation set to "true".
- A selector matching ScyllaDBDatacenters (scylla.scylladb.com/v1alpha1) with `scylla-operator.scylladb.com/register-with-global-scylladb-manager: "true"` label.

> The standalone manager-controller must no longer be deployed.

Author: > I think that we should adjust the design to ensure that existing setups will continue to work.

Existing setups will continue to work. Partially updated setups not necessarily. We should at least decide on the subset of partial updates that we want to support - I don't think we should support every possible misconfiguration.

> The standalone manager-controller must no longer be deployed.

Author: Addressed this in version skew / upgrade sections

)
```
where `FNV-1a` computes a base-36 string representation of a 64-bit FNV-1a hash. `SCYLLADB_CLUSTER_*` parameters refer to ScyllaDBDatacenter or ScyllaDBCluster objects. 

Scylla Operator will be extended with a controller reconciling ScyllaDBManagerClusterRegistration, responsible for registering ScyllaDBDatacenter and ScyllaDBCluster objects with a specified ScyllaDBManager instance.
For the time being, only the global ScyllaDBManager instance will be supported.

To mark ScyllaDBClusters or ScyllaDBDatacenters for registration with the global ScyllaDB Manager server, users will label the objects with a dedicated label: `"scylla-operator.scylladb.com/register-with-global-scylladb-manager": "true"`, matched by the selector of the global ScyllaDBManager instance.
To avoid ScyllaDB manager registering the same cluster through both ScyllaDBCluster and the underlying ScyllaDBDatacenters, the ScyllaDBCluster controller will not propagate the label to the created ScyllaDBDatacenters.

// TODO: names of clusters in manager state

#### Definition of `ScyllaDBManagerClusterRegistration.scylla.scylladb.com/v1alpha1`

```go

> type LocalScyllaDBReference struct {

Author: > suggestion: to avoid reinventing the wheel, use [corev1 ObjectReference](https://pkg.go.dev/k8s.io/api/core/v1#ObjectReference) and only allow a subset of fields to be used.

As per https://github.com/scylladb/scylla-operator/pull/2558#discussion_r2021340234, for local references this is what the developers suggest.

In case of the internal resource switching to corev1 ObjectReference with extra validation seems fine to me.

> type LocalScyllaDBReference struct {

Author: > Also the canonical way of referencing other kinds is by the entire GVK, not just the kind name. -> suggestion: replace Kind with the entire GVK.

https://github.com/kubernetes/kubernetes/blob/32cc146f75aad04beaaa245a7157eb35063a9f99/staging/src/k8s.io/api/core/v1/types.go#L6983-L6993

This is (almost) a verbatim copy of TypedLocalObjectReference, except the group is implied by the type name. What do we gain from GVK here?

	// kind specifies the type of the resource.
	Kind string `json:"kind"`
	// name specifies the name of the resource.
	Name string `json:"name"`
}

type ScyllaDBManagerClusterRegistrationSpec struct {
	// scyllaDBManagerRef specifies the reference to the local ScyllaDBManager that the cluster should be registered with.

> 	ScyllaDBManagerRef string `json:"scyllaDBManagerRef"`

Author: From k8s api LocalObjectReference definition:
https://github.com/kubernetes/kubernetes/blob/32cc146f75aad04beaaa245a7157eb35063a9f99/staging/src/k8s.io/api/core/v1/types.go#L6939-L6964
```
// New uses of this type are discouraged because of difficulty describing its usage when embedded in APIs.
...
// Instead of using this type, create a locally provided and used type that is well-focused on your reference.
```

Use of corev1 ObjectReference seems fine given it's an internal API so we don't make the UX worse for the users.

> 	ScyllaDBManagerRef string `json:"scyllaDBManagerRef"`

Reviewer: The goal in my mind is to have a (1) canonical, (2) open to extension, way of expressing references. The most likely extension coming to my mind is adding support for cross-namespace.

I wasn't aware of the upstream recommendation - that sounds good to me.

> Moving away from the string to an extensible type

sounds like a reasonable choice to me that addresses my concern.

> 	ScyllaDBManagerRef string `json:"scyllaDBManagerRef"`

Author: no longer applies

}
```

#### `ScyllaCluster.scylla.scylladb.com/v1` conversion mechanism

> The existing ScyllaCluster (scylla.scylladb.com/v1) controller, running as part of the standalone manager-controller, will be removed. Backup and repair tasks defined in `scyllaclusters.scylla.scylladb.com/v1.spec.backups` and `scyllaclusters.scylla.scylladb.com/v1.spec.repairs` will no longer be reconciled directly.

Reviewer: duplicate content?

https://github.com/scylladb/scylla-operator/pull/2558/commits/7e8c3eb9e3629ecbbb46540a856031d7e99f667d#diff-6eb3031404c03b7931ca976ae177107512a8cac3e65f6e5a4798aac4a8537a07R266-R268

Author: I don't see the duplication - one is about converting the spec into the new API, the other about backpropagating the status.
